st2common/st2common/util/param.py

```diff
@@ -170,17 +170,80 @@ def _process_defaults(G, schemas):
                 _process(G, name, value.get("default"))
 
 
+def _check_any_bad(G, nodes, check_any_bad=None, tracked_parents=None):
+    """
+    :param G: nx.DiGraph
+    :param nodes: list[dict]
+    :param check_any_bad: list[boolean]
+    :param tracked_parents: list[str]
+    """
+    if tracked_parents is None:
+        tracked_parents = []
+    if check_any_bad is None:
+        check_any_bad = []
+    for name in nodes:
+        if "value" not in G.nodes[name] and "template" not in G.nodes[name]:
+            # this is a string not a jinja template;  embedded {{sometext}}
+            check_any_bad.append(False)
+        else:
+            check_any_bad.append(True)
+        children = [i for i in G.predecessors(name)]
+        if name not in tracked_parents:
+            tracked_parents.extend(nodes)
+            _check_any_bad(G, children, check_any_bad, tracked_parents)
```

Comment:
```suggestion
        if name not in tracked_parents:
            tracked_parents.extend(nodes)
            children = [i for i in G.predecessors(name)]
            _check_any_bad(G, children, check_any_bad, tracked_parents)
```
Is there a reason to calculate children on every iteration? Or can we move it into this execute-once block?

Reply:
I am only processing the direct children here so it needs to happen for each iteration.  Also, the function is recursive to make the code simpler.

Reply:
oh i see what you meant. yes we could do it that way.